File: backend/app/schemas/plaza.py
"""题目广场相关的请求/响应模型"""

from datetime import datetime

from pydantic import BaseModel


class PlazaQuestionSetItem(BaseModel):
    """广场试题集列表项"""

    id: int
    title: str
    description: str | None = None
    teacher_name: str
    course_name: str
    shared_to_plaza_at: datetime
    attempt_count: int = 0
    average_score: float | None = None
    my_status: str | None = None  # null / "draft" / "completed"
    my_score: float | None = None


class PlazaQuestionSetListResponse(BaseModel):
    """广场试题集列表响应"""

    items: list[PlazaQuestionSetItem]
    total: int
    skip: int
    limit: int


class LeaderboardEntry(BaseModel):
    """排行榜条目"""

    rank: int
    user_name: str
    user_type: str  # "student" / "teacher"
    score: float
    submitted_at: datetime


class PlazaQuestionSetDetail(BaseModel):
    """广场试题集详情"""

    id: int
    title: str
    description: str | None = None
    teacher_name: str
    course_name: str
    shared_to_plaza_at: datetime
    attempt_count: int = 0
    completion_count: int = 0
    average_score: float | None = None
    my_status: str | None = None
    my_score: float | None = None
    my_rank: int | None = None
    created_at: datetime
    leaderboard: list[LeaderboardEntry] = []


class LeaderboardResponse(BaseModel):
    """排行榜响应"""

    question_set_id: int
    leaderboard: list[LeaderboardEntry]
    my_rank: int | None = None
    my_score: float | None = None


class PlazaAttemptItem(BaseModel):
    """我的广场练习列表项"""

    answer_id: int
    question_set_id: int
    question_set_title: str
    teacher_name: str
    status: str
    total_score: float | None = None
    submitted_at: datetime | None = None


class PlazaAttemptListResponse(BaseModel):
    """我的广场练习列表响应"""

    items: list[PlazaAttemptItem]
    total: int
    skip: int
    limit: int


class PlazaSharedStatItem(BaseModel):
    """教师分享到广场的试题集统计"""

    question_set_id: int
    title: str
    shared_to_plaza_at: datetime
    attempt_count: int = 0
    completion_count: int = 0
    average_score: float | None = None
    highest_score: float | None = None
    lowest_score: float | None = None


class PlazaSharedStatsResponse(BaseModel):
    """教师分享统计响应"""

    total_shared: int
    total_attempts: int
    items: list[PlazaSharedStatItem]


class SharePlazaResponse(BaseModel):
    """分享到广场响应"""

    message: str
    shared_to_plaza_at: datetime
    share_url: str


class UnsharePlazaResponse(BaseModel):
    """从广场撤回响应"""

    message: str


class PlazaAnswerSaveDraft(BaseModel):
    """教师做广场题保存草稿"""

    question_set_id: int
    student_answers: dict


class PlazaAnswerSubmit(BaseModel):
    """教师做广场题提交答案"""

    question_set_id: int
    student_answers: dict

File: frontend/src/services/plaza.ts
import api from './api';

export interface PlazaQuestionSetItem {
  id: number;
  title: string;
  description: string | null;
  teacher_name: string;
  course_name: string;
  shared_to_plaza_at: string;
  attempt_count: number;
  average_score: number | null;
  my_status: string | null; // null / "draft" / "completed"
  my_score: number | null;
  is_own: boolean;
}

export interface PlazaQuestionSetListResponse {
  items: PlazaQuestionSetItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface LeaderboardEntry {
  rank: number;
  user_name: string;
  user_type: string;
  score: number;
  submitted_at: string;
}

export interface PlazaQuestionSetDetail {
  id: number;
  title: string;
  description: string | null;
  teacher_name: string;
  course_name: string;
  shared_to_plaza_at: string;
  attempt_count: number;
  completion_count: number;
  average_score: number | null;
  my_status: string | null;
  my_score: number | null;
  my_rank: number | null;
  is_own: boolean;
  created_at: string;
  leaderboard: LeaderboardEntry[];
}

export interface LeaderboardResponse {
  question_set_id: number;
  leaderboard: LeaderboardEntry[];
  my_rank: number | null;
  my_score: number | null;
}

export interface PlazaAttemptItem {
  answer_id: number;
  question_set_id: number;
  question_set_title: string;
  teacher_name: string;
  status: string;
  total_score: number | null;
  submitted_at: string | null;
}

export interface PlazaAttemptListResponse {
  items: PlazaAttemptItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface PlazaSharedStatItem {
  question_set_id: number;
  title: string;
  shared_to_plaza_at: string;
  attempt_count: number;
  completion_count: number;
  average_score: number | null;
  highest_score: number | null;
  lowest_score: number | null;
}

export interface PlazaSharedStatsResponse {
  total_shared: number;
  total_attempts: number;
  items: PlazaSharedStatItem[];
}

export interface SharePlazaResponse {
  message: string;
  shared_to_plaza_at: string;
  share_url: string;
}


export const plazaApi = {
  // 广场列表（游客可访问）
  list: (params?: {
    skip?: number;
    limit?: number;
    keyword?: string;
    sort?: 'newest' | 'popular' | 'oldest';
  }) => api.get<PlazaQuestionSetListResponse>('/plaza/question-sets', { params }),

  // 广场详情（游客可访问）
  detail: (questionSetId: number) =>
    api.get<PlazaQuestionSetDetail>(`/plaza/question-sets/${questionSetId}`),

  // 排行榜
  leaderboard: (questionSetId: number, limit?: number) =>
    api.get<LeaderboardResponse>(`/plaza/question-sets/${questionSetId}/leaderboard`, {
      params: limit ? { limit } : {},
    }),

  // 题目内容（需登录）
  getContent: (questionSetId: number) =>
    api.get<{ id: number; title: string; json_content: string }>(
      `/plaza/question-sets/${questionSetId}/content`
    ),

  // 我的广场记录
  myAttempts: (params?: { skip?: number; limit?: number; status?: string }) =>
    api.get<PlazaAttemptListResponse>('/plaza/my-attempts', { params }),

  // 教师分享统计
  mySharedStats: () => api.get<PlazaSharedStatsResponse>('/plaza/my-shared-stats'),

  // 获取我的答案
  getMyAnswer: (questionSetId: number) =>
    api.get<{
      id: number;
      question_set_id: number;
      student_id: number | null;
      course_id: number;
      student_answers: Record<string, string | string[]> | null;
      grading_results: Record<string, unknown> | null;
      total_score: number | null;
      status: string;
      error_message: string | null;
      saved_at: string;
      submitted_at: string | null;
    } | null>(`/plaza/my-answer/${questionSetId}`),

  // 做题 - 保存草稿
  saveDraft: (data: { question_set_id: number; student_answers: Record<string, unknown> }) =>
    api.post('/plaza/answer/save-draft', data),

  // 做题 - 提交答案
  submitAnswer: (data: { question_set_id: number; student_answers: Record<string, unknown> }) =>
    api.post('/plaza/answer/submit', data),

  // 分享到广场（教师端，在 question API 下）
  sharePlaza: (questionSetId: number) =>
    api.post<SharePlazaResponse>(`/question/${questionSetId}/share-plaza`),

  // 从广场撤回（教师端）
  unsharePlaza: (questionSetId: number) =>
    api.post<{ message: string }>(`/question/${questionSetId}/unshare-plaza`),
};
